### ISSUE TITLE
ioImplementation thing should provide interface for getting files.

### ISSUE THREAD

**Issue by christian-marie** [anchor/git-vogue#59]
Grabbing all or changed files is currently hidden in the IOs of the  plugin runner implementation. Exposing this as a separate function would allow us to do all kinds of smarts generically. This is a good way of achieving #54 and generally cleaning up the codebase/paving the way for future improvements.

**Comment by thsutton** [comment]
Incidentally, `git ls-files` has an argument for an ignore file to use; so we can do the additional "don't validate these plox" file. Alas, there doesn't seem to be an analogous argument for `git diff`.

**Comment by christian-marie** [comment]
I have started work on this in the wip-modular-backends branch.

**Comment by christian-marie** [comment]
Relevant types are:

```haskell
-- | A plugin that can be told to check or fix a list of files
data Plugin m = Plugin
    { pluginName :: Text
    , runCheck   :: [FilePath] -> m (Status Check)
    , runFix     :: [FilePath] -> m (Status Fix)
    }

-- | An implementation of a "runner" of plugins. Mostly for easy testing.
data PluginExecutorImpl m = PluginExecutorImpl{
    executeFix   :: [FilePath] -> Plugin m -> m (Status Fix),
    executeCheck :: [FilePath] -> Plugin m -> m (Status Check)
}

-- | A thing that can find plugins, for example we might search through the
-- libexec directory for executables.
data PluginDiscoverer m = PluginDiscoverer {
    discoverPlugins :: m [Plugin m]
}

-- | A VCS backend, such as git.
data VCS m = VCS
    { getConfig    :: Text -> m Text             -- ^ Get the associated config
    , setConfig    :: Text -> Text -> m ()       -- ^ Set associated config
    , getFiles     :: SearchMode -> m [FilePath] -- ^ Find all staged files
    , installHook  :: m ()                       -- ^ Install pre-commit hook
    , removeHook   :: m ()                       -- ^ Remove pre-commit hook
    , checkHook    :: m Bool                     -- ^ Check pre-commit hook
    , findTopLevel :: m FilePath                 -- ^ Find the / of the repo
    }
```

**Comment by thsutton** [comment]
`getConfig` should be `m (Maybe Text)` or similar unless we want to restrict `m` to be a `MonadError`?

**Comment by christian-marie** [comment]
Fixed by #62
